eclair-core/src/main/scala/fr/acinq/eclair/channel/InteractiveTxBuilder.scala

```diff
@@ -0,0 +1,722 @@
+                                   channelFlags: ChannelFlags,
+                                   channelConfig: ChannelConfig,
+                                   channelFeatures: ChannelFeatures,
+                                   wallet: OnChainChannelFunder,
+                                   previousAttempts: Seq[InteractiveTxBuilder.SignedSharedTransaction],
+                                   timers: TimerScheduler[InteractiveTxBuilder.Command],
+                                   context: ActorContext[InteractiveTxBuilder.Command])(implicit ec: ExecutionContext) {
+
+  import InteractiveTxBuilder._
+
+  private val log = context.log
+
+  def start(): Behavior[Command] = {
+    val toFund = if (fundingParams.isInitiator) {
+      // If we're the initiator, we need to pay the fees of the common fields of the transaction, even if we don't want
+      // to contribute to the shared output.
+      fundingParams.localAmount.max(fundingParams.dustLimit)
+    } else {
+      fundingParams.localAmount
+    }
+    log.debug("contributing {} to interactive-tx construction", toFund)
+    if (toFund <= 0.sat) {
+      // We're not the initiator and we don't want to contribute to the funding transaction.
+      buildTx(FundingContributions(Nil, Nil))
+    } else {
+      // We always double-spend all our previous inputs.
+      val previousInputs = previousAttempts.flatMap(_.tx.localInputs).distinctBy(_.serialId)
+      val dummyTx = Transaction(2, previousInputs.map(i => TxIn(toOutPoint(i), ByteVector.empty, i.sequence)), Seq(TxOut(toFund, fundingParams.fundingPubkeyScript)), fundingParams.lockTime)
+      fund(dummyTx, previousInputs, Set.empty)
+    }
+  }
+
+  def fund(txNotFunded: Transaction, currentInputs: Seq[TxAddInput], unusableInputs: Set[OutPoint]): Behavior[Command] = {
+    context.pipeToSelf(wallet.fundTransaction(txNotFunded, fundingParams.targetFeerate, replaceable = true, lockUtxos = true)) {
+      case Failure(t) => WalletFailure(t)
+      case Success(result) => FundTransactionResult(result.tx)
+    }
+    Behaviors.receiveMessagePartial {
+      case FundTransactionResult(fundedTx) =>
+        filterInputs(fundedTx, currentInputs, unusableInputs)
+      case WalletFailure(t) =>
+        log.error("could not fund dual-funded channel: ", t)
+        // We use a generic exception and don't send the internal error to the peer.
+        replyTo ! LocalFailure(ChannelFundingError(fundingParams.channelId))
+        unlockAndStop(currentInputs.map(toOutPoint).toSet ++ unusableInputs)
+      case msg: ReceiveMessage =>
+        timers.startSingleTimer(msg, 1 second)
+        Behaviors.same
+      case Abort =>
+        timers.startSingleTimer(Abort, 1 second)
+        Behaviors.same
+    }
+  }
+
+  def filterInputs(fundedTx: Transaction, currentInputs: Seq[TxAddInput], unusableInputs: Set[OutPoint]): Behavior[Command] = {
+    context.pipeToSelf(Future.sequence(fundedTx.txIn.map(txIn => getInputDetails(txIn, currentInputs)))) {
```

Comment:
Same way we use `currentInputs` to not make superfluous bitcoind calls, we could also use `unusableInputs`:

```suggestion
    context.pipeToSelf(Future.sequence(fundedTx.txIn
      .filterNot(txIn => unusableInputs.contains(txIn.outPoint))
      .map(txIn => getInputDetails(txIn, currentInputs)))) {
```

Reply:
In theory this should be useless: we keep the unusable inputs locked explicitly to prevent `bitcoind` from re-adding them when funding (otherwise we'd be in an infinite funding loop). Maybe if there is an intersection between `fundedTx.txIn` and `unusableInputs`, we should just abort instead of risking going into this infinite loop? I'll investigate that.

Reply:
The assymetry is what I'm after, more than the optimization.

Reply:
I've fixed this in https://github.com/ACINQ/eclair/pull/2273/commits/6e0c15d430e3f4867e97196e7e350701753db688
I don't think we should strive for symmetry here, these are different things.

Reply:
> Maybe if there is an intersection between fundedTx.txIn and unusableInputs, we should just abort instead of risking going into this infinite loop? I'll investigate that.

Good idea, fix LGTM